flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/InternalTimerHeapTest.java

```diff
@@ -0,0 +1,470 @@
+		Assert.assertFalse(timerPriorityQueue.stopTimer(timestamp, key, VoidNamespace.INSTANCE));
+		Assert.assertTrue(timerPriorityQueue.isEmpty());
+	}
+
+	private void testAddOfferCommon(
+		BiFunction<InternalTimerHeap, TimerHeapInternalTimer<Integer, VoidNamespace>, Boolean> testMethod) {
+
+		InternalTimerHeap<Integer, VoidNamespace> timerPriorityQueue = newPriorityQueue(1);
+
+		Assert.assertTrue(timerPriorityQueue.isEmpty());
+
+		TimerHeapInternalTimer<Integer, VoidNamespace> timer1 =
+			new TimerHeapInternalTimer<>(42L, 4711, VoidNamespace.INSTANCE);
+		TimerHeapInternalTimer<Integer, VoidNamespace> timer2 =
+			new TimerHeapInternalTimer<>(43L, 4712, VoidNamespace.INSTANCE);
+
+		// add first timer
+		Assert.assertTrue(testMethod.apply(timerPriorityQueue, timer1));
+		Assert.assertEquals(1, timerPriorityQueue.size());
+
+		// add second timer
+		Assert.assertTrue(testMethod.apply(timerPriorityQueue, timer2));
+		Assert.assertEquals(2, timerPriorityQueue.size());
+
+		// check adding first timer again, duplicate should not be added
+		Assert.assertFalse(testMethod.apply(timerPriorityQueue, timer1));
+		Assert.assertEquals(2, timerPriorityQueue.size());
+	}
+
+	private void testPeekElementCommon(
+		InternalTimerHeap<Integer, VoidNamespace> timerPriorityQueue,
+		Function<InternalTimerHeap, TimerHeapInternalTimer<Integer, VoidNamespace>> testFun) {
+
+		TimerHeapInternalTimer<Integer, VoidNamespace> timer1 =
+			new TimerHeapInternalTimer<>(42L, 4711, VoidNamespace.INSTANCE);
+		timerPriorityQueue.add(timer1);
+
+		Assert.assertEquals(timer1, testFun.apply(timerPriorityQueue));
+
+		TimerHeapInternalTimer<Integer, VoidNamespace> timer2 =
+			new TimerHeapInternalTimer<>(43L, 4712, VoidNamespace.INSTANCE);
+		timerPriorityQueue.add(timer2);
+
+		Assert.assertEquals(timer1, testFun.apply(timerPriorityQueue));
+
+		TimerHeapInternalTimer<Integer, VoidNamespace> timer3 =
+			new TimerHeapInternalTimer<>(41L, 4712, VoidNamespace.INSTANCE);
+		timerPriorityQueue.add(timer3);
+
+		Assert.assertEquals(timer3, testFun.apply(timerPriorityQueue));
+		Assert.assertEquals(3, timerPriorityQueue.size());
+	}
+
+	private void testRemovePollCommon(
+		InternalTimerHeap<Integer, VoidNamespace> timerPriorityQueue,
+		Function<InternalTimerHeap<Integer, VoidNamespace>, TimerHeapInternalTimer<Integer, VoidNamespace>> fun) {
+
+		final int testSize = 345;
+		HashSet<TimerHeapInternalTimer<Integer, VoidNamespace>> checkSet = new HashSet<>(testSize);
+		insertRandomTimers(timerPriorityQueue, checkSet, testSize);
+
+		long lastTimestamp = Long.MIN_VALUE;
+		while (!timerPriorityQueue.isEmpty()) {
+			TimerHeapInternalTimer<Integer, VoidNamespace> removed = fun.apply(timerPriorityQueue);
+			Assert.assertTrue(checkSet.remove(removed));
+			Assert.assertTrue(removed.getTimestamp() >= lastTimestamp);
+			lastTimestamp = removed.getTimestamp();
+		}
+		Assert.assertTrue(checkSet.isEmpty());
+	}
```

Comment:
Do we test somewhere the serialization and deserialization of the `TimerHeapInternalTimer`?